src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -89,8 +89,34 @@ void GcsNodeManager::NodeFailureDetector::DetectDeadNodes() {
 void GcsNodeManager::NodeFailureDetector::SendBatchedHeartbeat() {
   if (!heartbeat_buffer_.empty()) {
     auto batch = std::make_shared<rpc::HeartbeatBatchTableData>();
-    for (const auto &heartbeat : heartbeat_buffer_) {
-      batch->add_batch()->CopyFrom(heartbeat.second);
+    std::unordered_map<SchedulingClass, rpc::ResourceDemand> aggregate_load;
+    for (auto &heartbeat : heartbeat_buffer_) {
+      // Aggregate the load reported by each raylet.
+      auto load = heartbeat.second.resource_load_by_shape();
+      for (const auto &demand : load.resource_demands()) {
+        auto scheduling_class =
+            TaskSpecification::GetSchedulingClass(MapFromProtobuf(demand.shape()));
```

Comment:
It's a little odd to use the scheduling class a temporary key here, I guess there isn't any other convenient key function that would work?

Reply:
I replaced this with the `ResourceSet` class.